File: Java/213. House Robber II.java
class Solution {
    public int rob(int[] nums) {        
        int pre = 0, curr = 0, max = 0;
        for (int i = 0; i < nums.length - 1; i++) {
            int tmp = curr;
            curr = Math.max(pre + nums[i], curr);
            pre = tmp;
        }
        max = curr;        
        pre = curr = 0;
        for (int i = 1; i < nums.length; i++) {
            int tmp = curr;
            curr = Math.max(pre + nums[i], curr);
            pre = tmp;
        }
        max = Math.max(max, curr);
        if (nums.length > 0)            
            return Math.max(max, nums[0]);
        else
            return max;
    }
}

File: Python/213. House Robber II.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        prev = curr = 0
        for num in nums[1:]:
            tmp = curr
            curr = max(curr, prev + num)
            prev = tmp
        ans = curr
        prev = curr = 0
        for num in nums[:-1]:
            tmp = curr
            curr = max(curr, prev + num)
            prev = tmp
        return nums[0] if len(nums) == 1 else max(curr, ans)
